crates/autopilot/src/on_settlement_event_updater.rs

```diff
@@ -0,0 +1,342 @@
+// transaction and insert them into the database (settlement_observations
+// table). Now we know which tokens were used in the transaction so we
+// update the auction_prices table with the actual prices that were used.
+
+use {
+    crate::{
+        database::{on_settlement_event_updater::SettlementUpdate, Postgres},
+        decoded_settlement::{DecodedSettlement, FeeConfiguration},
+    },
+    anyhow::{anyhow, Context, Result},
+    contracts::GPv2Settlement,
+    num::BigRational,
+    primitive_types::{H160, H256},
+    shared::{
+        current_block::CurrentBlockStream,
+        ethrpc::Web3,
+        event_handling::MAX_REORG_BLOCK_COUNT,
+        external_prices::ExternalPrices,
+    },
+    std::{collections::HashSet, time::Duration},
+    web3::types::TransactionId,
+};
+
+pub struct OnSettlementEventUpdater {
+    pub web3: Web3,
+    pub contract: GPv2Settlement,
+    pub native_token: H160,
+    pub db: Postgres,
+    pub current_block: CurrentBlockStream,
+    pub fee_objective_scaling_factor: BigRational,
+}
+
+impl OnSettlementEventUpdater {
+    pub async fn run_forever(self) -> ! {
+        loop {
+            match self.update().await {
+                Ok(true) => (),
+                Ok(false) => tokio::time::sleep(Duration::from_secs(10)).await,
+                Err(err) => {
+                    tracing::error!(?err, "on settlement event update task failed");
+                    tokio::time::sleep(Duration::from_secs(10)).await;
+                }
+            }
+        }
+    }
+
+    /// Update database for settlement events that have not been processed yet.
+    ///
+    /// Returns whether an update was performed.
+    async fn update(&self) -> Result<bool> {
+        let current_block = self.current_block.borrow().number;
+        let reorg_safe_block: i64 = current_block
+            .checked_sub(MAX_REORG_BLOCK_COUNT)
+            .context("no reorg safe block")?
+            .try_into()
+            .context("convert block")?;
+
+        let event = match self
+            .db
+            .get_settlement_event_without_tx_info(reorg_safe_block)
+            .await
+            .context("get_settlement_event_without_tx_info")?
+        {
+            Some(event) => event,
+            None => return Ok(false),
+        };
+
+        let hash = H256(event.tx_hash.0);
+        tracing::trace!("updating settlement details for tx {hash:?}");
+
+        let transaction = self
+            .web3
+            .eth()
+            .transaction(TransactionId::Hash(hash))
+            .await
+            .with_context(|| format!("get tx {hash:?}"))?
+            .with_context(|| format!("no tx {hash:?}"))?;
+        let tx_from = transaction
+            .from
+            .with_context(|| format!("no from {hash:?}"))?;
+        let tx_nonce: i64 = transaction
+            .nonce
+            .try_into()
+            .map_err(|err| anyhow!("{}", err))
+            .with_context(|| format!("convert nonce {hash:?}"))?;
```

Comment:
Are these not available in the transaction receipt?

Reply:
is the transaction not the same as transaction receipt?

Reply:
No, they are not the same. In this code we are fetching both, and just wanted to double check it is really needed.

Reply:
They should exist but I wouldn't rely on it since the fields are `Option<T>`